models/Version/Adapter/DelegateVersionStorageAdapter.php

```diff
@@ -0,0 +1,134 @@
+ * - Pimcore Commercial License (PCL)
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Model\Version\Adapter;
+
+class DelegateVersionStorageAdapter implements VersionStorageAdapterInterface
+{
+    public function __construct(protected array $adapters,
+                                protected int $byte_threshold,
+                                protected VersionStorageAdapterInterface $defaultAdapter,
+                                protected VersionStorageAdapterInterface $fallbackAdapter)
+    {
+
+    }
+
+    protected function getAdapter(string $storageType = null): VersionStorageAdapterInterface
+    {
+        if(empty($storageType) === true) {
+            return $this->defaultAdapter;
+        }
+        else {
+            $adapter = $this->adapters[$storageType] ?? null;
+        }
+        if(isset($adapter) === false || isset($adapter['class']) === false) {
+            throw new \Exception("no adapter for storage type " . $storageType . " found.");
+        }
+        return $adapter['class'];
+    }
+
+    public function loadMetaData(int $id,
+                                 int $cId,
+                                 string $cType,
+                                 string $storageType = null) : ?string {
+
+        $storageType = $storageType ?? $this->defaultAdapter;
+        return $this->getAdapter($storageType)->loadMetaData($id,
+                                                            $cId,
+                                                            $cType,
+                                                            $storageType);
+    }
+
+    public function loadBinaryData(int $id,
+                                   int $cId,
+                                   string $cType,
+                                   string $storageType = null,
+                                   int $binaryFileId = null): mixed
+    {
+        $storageType = $storageType ?? $this->defaultAdapter;
+        return $this->getAdapter($storageType)->loadBinaryData($id,
+                                                                $cId,
+                                                                $cType,
+                                                                $storageType,
+                                                                $binaryFileId);
+    }
+
+    public function getStorageType(string $metaData,
+                                   mixed $binaryDataStream = null): string {
+        if(empty($this->fallbackAdapter) === false) {
+            $binarySize = 0;
+            $metaDataSize = strlen($metaData);
+
+            if (isset($binaryDataStream) === true) {
+                $stats = fstat($binaryDataStream);
+                $binarySize = $stats['size'];
```

Comment:
I'm not sure, whether `size` is available for all types of streams (e.g. S3). 
So I'd go for the storage path here and use `$storage->fileSize($path)` instead.